feat: add examples page

cdrage opened the pull request:
feat: add examples page

### What does this PR do?

* Minor edit to category (see json update)
* Adds an example page (rendered portion) that shows all the examples we
  currently have
* Pull and build image buttons
* Reactively do in-progress for pulling

### Screenshot / video of UI

<!-- If this PR is changing UI, please include
screenshots or screencasts showing the difference -->


https://github.com/user-attachments/assets/9a9d67a3-ecc8-4367-aea9-1061d2fa81de

![Screenshot 2024-10-17 at 3 34 15 PM](https://github.com/user-attachments/assets/eb136041-9a9e-49e0-b5c1-07bb60a186ae)
![Screenshot 2024-10-17 at 3 34 25 PM](https://github.com/user-attachments/assets/03fc7f1e-6651-41c4-bdf3-43b717d642c1)



### What issues does this PR fix or reference?

<!-- Include any related issues from Podman Desktop
repository (or from another issue tracker). -->

Closes https://github.com/containers/podman-desktop-extension-bootc/issues/895

### How to test this PR?

<!-- Please explain steps to reproduce -->

1. Go to examples page
2. Press pull image button
3. Then build image button after it finishes

Signed-off-by: Charlie Drage <charlie@charliedrage.com>

cdrage commented:
Tests have been added! Ready for testing out and using 🚀 

I have added some "follow up issues" to the topic here, especially lack of examples (only 3 "good" ones right now): https://github.com/containers/podman-desktop-extension-bootc/issues/864#issuecomment-2420381794

benoitf commented on packages/frontend/src/lib/ExamplesCard.svelte:
```diff
@@ -0,0 +1,68 @@
+      const example = examples.find(example => example.image === msg.image);
+      if (example) {
+        example.state = 'pulled';
+        console.log(examples);
```
to be removed

benoitf commented on packages/frontend/src/lib/ExamplesCard.svelte:
```diff
@@ -0,0 +1,68 @@
+// Function to update examples based on available images
+function updateExamplesWithPulledImages() {
+  if (bootcAvailableImages) {
+    bootcAvailableImages.forEach(image => {
```
I think the linter is now recommending. for of pattern

benoitf commented on packages/frontend/src/lib/ExamplesCard.svelte:
```diff
@@ -0,0 +1,68 @@
+    bootcAvailableImages.forEach(image => {
+      // Only do it if there is a RepoTags
+      if (image.RepoTags) {
+        const [imageRepo, imageTag] = image.RepoTags[0].split(':');
```
maybe use the optional ? pattern to reduce the if if if chaining

benoitf commented on packages/frontend/src/lib/ExamplesCard.svelte:
```diff
@@ -0,0 +1,68 @@
+}
+
+// Reactive statement to call the function each time bootcAvailableImages updates
+$: if (bootcAvailableImages) {
```
might be a time to use $effect / svelte 5 runes

cdrage replied:
Can't as we use `$$props.class` in the Card (copied from AI lab and we try to keep consistent with them too) which we can then not use since it puts the svelte file in "runes mode"

cdrage replied:
NVM, we don't actually use classes imported into it, so it's okay to switch to svelte 5 runes, updating now.

benoitf replied:
I don't see any reason why it would fail

benoitf commented on packages/frontend/src/lib/ExamplesCard.svelte:
```diff
@@ -0,0 +1,68 @@
+<Card title={category.name} classes="{$$props.class} font-medium mt-4">
+  <div slot="content" class="w-full">
+    {#if examples.length === 0}
+      <div class="text-gray-400 mt-2">There is no example in this category.</div>
```
should ot use empty screen ?

cdrage replied:
No need :) will always have examples! 

Tried your suggestion though using empty screen but messed up formatting, so may be easier to just use this.

benoitf replied:
todo: if we always have examples, then do not add the if clause

benoitf commented on packages/frontend/src/lib/ExamplesCard.svelte:
```diff
@@ -0,0 +1,68 @@
+<script lang="ts">
+import Card from './Card.svelte';
+import ExampleCard from './ExampleCard.svelte';
+import type { Example, Category } from '../../../shared/src/models/examples';
```
might replace all the ../.. by `@something/`

benoitf commented on packages/frontend/src/lib/ExampleCard.spec.ts:
```diff
@@ -0,0 +1,115 @@
+import { render, fireEvent, screen } from '@testing-library/svelte';
+import { expect, test, vi } from 'vitest';
+import ExampleCard from './ExampleCard.svelte';
+import { bootcClient } from '../api/client';
```
use absolute import

benoitf commented on packages/frontend/src/lib/Card.svelte:
```diff
@@ -0,0 +1,33 @@
+<script lang="ts">
```
might switch to $props / runes

benoitf commented on packages/frontend/src/Examples.svelte:
```diff
@@ -0,0 +1,56 @@
+
+onMount(async () => {
+  // onmount get the examples
+  let examples = await bootcClient.getExamples();
```
it means that it's only when we create the component, it won't be reactive it's ok ?

cdrage replied:
okay to not be reactive as we as not making any changes to the assets file reactively / modifying it / or getting it from a local file.

benoitf commented on packages/backend/assets/examples.json:
```diff
@@ -38,7 +38,7 @@
     {
       "id": "fedora",
-      "name": "Fedora BootC Images"
+      "name": "Fedora Bootable Container Images"
```
could be a separate small PR

cdrage replied:
ACK

deboer-tim reviewed (commented):
Ditto to all the points @benoitf raised.

Worked fine when testing. The biggest gap to me was how to find details or instructions for each example. Why does it exist, what does it do, and how do I use it? The source goes to the repo root, then I have to guess at the folder and look at the readme. (and ok, Tailscale doesn't have anything)

Could we link directly to the folder, or ideally have a way to view the readme directly? Maybe needs a details page for each one?

Noted a few other minor usability issues:
- The page appears, _then_ we switch buttons from Pull to Build. Is the API fast enough that we could do this before opening the page, or can we delay adding the button until we know which it is? (might need to move Source button to Link on the left so it doesn't move).
- Like on the dashboard, you don't know the image you're about to pull is huge. Would be nice to include the size on the card, or warn the user when they click on the button.
- No pull task or progress. Unrelated to this issue so I opened up #944.

cdrage commented:
> Ditto to all the points @benoitf raised.
> 
> Worked fine when testing. The biggest gap to me was how to find details or instructions for each example. Why does it exist, what does it do, and how do I use it? The source goes to the repo root, then I have to guess at the folder and look at the readme. (and ok, Tailscale doesn't have anything)
> 
> Could we link directly to the folder, or ideally have a way to view the readme directly? Maybe needs a details page for each one?
> 
> Noted a few other minor usability issues:
> 
> * The page appears, _then_ we switch buttons from Pull to Build. Is the API fast enough that we could do this before opening the page, or can we delay adding the button until we know which it is? (might need to move Source button to Link on the left so it doesn't move).
> * Like on the dashboard, you don't know the image you're about to pull is huge. Would be nice to include the size on the card, or warn the user when they click on the button.
> * No pull task or progress. Unrelated to this issue so I opened up [Usability of pulling images #944](https://github.com/containers/podman-desktop-extension-bootc/issues/944).

We have https://gitlab.com/fedora/bootc/examples/-/issues/19 which I could work on next sprint with regards to descriptions of what each one does.

We have no way of knowing the size of the image before pulling, but we can at least add an approximate "size" on the description when pulling?

ACK on button delay / showing it correctly.

cdrage commented:
@deboer-tim @benoitf

Updated the code based on the review:

- Added size of the image (hardcoded in examples.json), makes it easier and less calls to check for size (honestly sizes will not change much, and this is just for showing examples)
- Modified how we show the buttons on load, less flickering / changes.
- Did not update to svelte5 (could not get it to work very well without causing extra changes in other svelte files)
- Updated rest of minor code edits

cdrage commented:
> I'm no longer getting Pull buttons when I don't have the image locally: <img alt="Screenshot 2024-10-21 at 5 03 30 PM" width="988" src="https://private-user-images.githubusercontent.com/19958075/378597762-dd1dbed2-ee48-4463-afb3-90c12c004e57.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjk2MDcxOTAsIm5iZiI6MTcyOTYwNjg5MCwicGF0aCI6Ii8xOTk1ODA3NS8zNzg1OTc3NjItZGQxZGJlZDItZWU0OC00NDYzLWFmYjMtOTBjMTJjMDA0ZTU3LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEwMjIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMDIyVDE0MjEzMFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTI2ODliNGMxYWE3MGRlYTk4YzNlYmJjYjhlMTYzNGNjZWZhZjEwYzc3MTYwNTA3YmQzY2U2Y2QyNzhmYzMyMzkmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.gwpeTZscKlgfnB4x1kLuqxk4jGFjE9SzEU-xxooajmk">

that's my mistake, I had messed up an `if` statement.

Latest commit I have also updated the styling, adding a "querying" button in case it takes a while to get the list of images, etc, as well as reduce the amount of flickering on initial load.

deboer-tim commented on packages/frontend/src/lib/Card.svelte:
```diff
@@ -0,0 +1,33 @@
+</script>
+
+<a class="no-underline" href={href}>
+  <div class="{classes} rounded-md flex-nowrap overflow-hidden" role="region" aria-label={title ?? 'Card'}>
```
We are doing this in a few places, often with different parameter names. In Svelte 5 you can do:
```
let {
  class: className = '',
}: {
  class?: string;
} = $props();
```
to make this cleaner, but since we ever only ever do `classes="font-medium mt-4"`, maybe remove the parameter and put the styling in here?

deboer-tim commented on packages/frontend/src/lib/ExampleCards.spec.ts:
```diff
@@ -0,0 +1,125 @@
+import { render, screen } from '@testing-library/svelte';
+import { expect, test, vi } from 'vitest';
+import type { Example, Category } from '/@shared/src/models/examples';
+import ExamplesCard from './ExamplesCard.svelte';
```
component: ExamplesCard, test: ExampleCards. Can you rename one so that they match?

deboer-tim commented on packages/frontend/src/Examples.svelte:
```diff
@@ -0,0 +1,56 @@
+<script lang="ts">
+import type { Example, Category } from '../../shared/src/models/examples';
```
Should really use `/@shared/...` here, and other places.

deboer-tim reviewed (commented):
Thanks for the changes, worked well in testing. We have some ideas for additional improvements open, but this delivers on the core requirement.

Some minor requests for change but otherwise this is good to go for me.

deboer-tim reviewed (approved):
Thanks for the changes @cdrage, a nice little addition we can continue to expand on.

benoitf commented on packages/frontend/src/Examples.spec.ts:
```diff
@@ -0,0 +1,140 @@
+
+  // Wait until example1 says Build image as it updates reactively
+  // same for example 2 but Pull image
+  waitFor(() => {
```
issue: missing await before waitFor

benoitf commented on packages/frontend/src/Examples.spec.ts:
```diff
@@ -0,0 +1,140 @@
+    expect(buildImage1).toBeInTheDocument();
+  });
+
+  waitFor(() => {
```
issue: missing await before waitFor

benoitf commented on packages/frontend/src/lib/Card.svelte:
```diff
@@ -0,0 +1,32 @@
+<script lang="ts">
+export let title: string | undefined = undefined;
+export let description: string | undefined = undefined;
+
+export let href: string | undefined = undefined;
```
suggestion: use svelte5 runes

```suggestion
interface Props {
title: string;
description: string;
href: string
}

let {title, description, href} : Props = $props();
```

benoitf commented on packages/frontend/src/lib/Card.svelte:
```diff
@@ -0,0 +1,32 @@
+export let href: string | undefined = undefined;
+</script>
+
+<a class="no-underline" href={href}>
```
```suggestion
<a class="no-underline" {href}>
```

benoitf commented on packages/frontend/src/lib/ExampleCard.svelte:
```diff
@@ -0,0 +1,111 @@
+  }
+}
+
+// Function that takes something in MB and returns it in GB is more than 1GB
+// in a nice string format
+function formatStringSize(size: number): string {
+  if (size > 1000) {
+    return `${(size / 1000).toFixed(1)} GB`;
+  }
+  return `${size} MB`;
+}
```
todo: please use humansize library

https://www.npmjs.com/package/human-size

cdrage replied:
I decided to use the filesize library https://www.npmjs.com/package/filesize which is used in podman desktop. the one you linked hasn't been updated in 10 years.

benoitf replied:
ah yes I think I took the wrong name but you figured it out
I mixed humanize-duration and this one

benoitf commented on packages/frontend/src/lib/ExampleCard.svelte:
```diff
@@ -0,0 +1,111 @@
+}
+
+// Make sure if example.pulled is updated, we force a re-render
+$: {
+  example.state;
+}
```
issue: feels hacky

use svelte5 ($props, $effect, etc)

benoitf commented on packages/frontend/src/lib/ExampleCard.svelte:
```diff
@@ -0,0 +1,111 @@
+          icon={faArrowDown}
+          aria-label="Pull image"
+          title="Pull image"
+          class="w-28"
```
question: do we need to set a size there ?

benoitf commented on packages/frontend/src/lib/ExampleCard.svelte:
```diff
@@ -0,0 +1,111 @@
+          inProgress={pullInProgress}>Pull image</Button>
+      {:else}
+        <!-- Show a spinner / in progress for querying button instead if we are still loading information-->
+        <Button icon={faArrowDown} aria-label="Querying" title="Querying" class="w-28" inProgress={true}
```
question: same here, why do we need to set a size there ? (w-28)

cdrage commented:
Updated @benoitf @benoitf 

The more I use svelte5, the more I'm liking it! Much cleaner code.

I have updated it to svelte5 paradigms. I also decided to use https://www.npmjs.com/package/filesize library instead of humansize since it's the same on in Podman Desktop + is more updated.

benoitf commented:
yes the library name was wrongly spelled !

cdrage commented:
> looks like there are some typechecking issue but approving

Thanks! Formatter had messed up and did not put `{}` around href. Added and updated now.
